File: bcw5/php/Unit/Unit.php
<?php
	class Unit {
        private $damage;
        private $hitPoints;
        private $hitPointsLimit;
        private $name;
        private function ensureIsAlive() {
        	if ( $this->hitPoints == 0 ) {
		        throw new Exception("Unit is Dead!");
		    }
        }

        public function Unit($name, $hp, $dmg) {
        	$this->damage = $dmg;
		    $this->hitPoints = $hp;
		    $this->hitPointsLimit = $hp;
		    $this->name = $name;
        }

        public function getDamage() {
        	$this->damage;
        }

        public function getHitPoints() {
        	$this->hitPoints;
        }

        public function getHitPointsLimit() {
        	$this->hitPointsLimit;
        }

        public function getName() {
        	$this->name;
        }

        public function addHitPoints($hp) {
        	$this->ensureIsAlive();

		    $healedHp = $this->hitPoints + $hp;

		    if ( $healedHp >  $this->hitPointsLimit ) {
		        $healedHp = $this->hitPointsLimit;
		    }
		    $this->hitPoints = $healedHp;
        }

        public function takeDamage($dmg) {
		    $this->ensureIsAlive();

		    $this->hitPoints -= $dmg;

		    if ( $this->hitPoints < 0 ) {
		        $this->hitPoints = 0;
		        throw new Exception("Unit is Dead!");
		    }
        }

        public function attack($enemy) {
        	$enemy->takeDamage($this->damage);
    		$enemy->counterAttack($this);
        }

        public function counterAttack($enemy) {
        	$enemy->takeDamage($this->damage/2);
        }

        public function output() {
        	echo PHP_EOL . "==================" . PHP_EOL .
        	"Unit name: " . $this->name . "," . PHP_EOL .
        	"Unit HP: (" . $this->hitPoints . "/" . $this->hitPointsLimit . ")" . "," .PHP_EOL .
        	"Unit damage: " . $this->damage . ";" . PHP_EOL . "==================" . PHP_EOL;
        }
	}
?>

File: bcw4/java/Gun/Unit.java
public class Unit {
    private int damage;
    private int hitPoints;
    private int hitPointsLimit;
    private String name;
    private void ensureIsAlive();

    public Unit(String name, int hp, int dmg) {
        this.name = name;
        this.hitPointsLimit = hp;
        this.hitPoints = hp;
        this.damage = dmg;
    }

    public Unit() {
        this.name = "Unit";
        this.hitPointsLimit = 10;
        this.hitPoints = 10;
        this.damage = 1;
    }

    public int getDamage() {
        return damage;
    }

    public int getHitPoints() {
        return hitPoints;
    }

    public int getHitPointsLimit() {
        return hitPointsLimit;
    }
    
    public String getName() {
        return name;
    }

    public void ensureIsAlive() {
        if ( hitPoints == 0 ) {
            throw UnitIsDeadException();
        }
    }

    public void addHitPoints(int hp) {
        ensureIsAlive();

        int healedHp = hitPoints + hp;

        if ( healedHp >  hitPointsLimit ) {
            healedHp = hitPointsLimit;
        }
        hitPoints = healedHp;
    }
    
    public void takeDamage(int dmg) {
        ensureIsAlive();

        hitPoints -= dmg;

        if ( hitPoints < 0 ) {
            hitPoints = 0;
            throw UnitIsDead();
        }
    }

    public void attack(Unit enemy) {
        enemy.takeDamage(damage);
        enemy.counterAttack(*this);
    }

    public void counterAttack(Unit enemy) {
        enemy.takeDamage(damage/2);
    }

    public String toString() {
        return "=========================\n" + "Unit name: " + getName() + "\nUnit hitpoints: " + getHitPoints() + "/" + getHitPointsLimit() + "\nUnit damage: " + getDamage() + "\n=========================";
    }
}